### ISSUE TITLE
grep unpack a/(a) loops and eats all memory

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9481]
Migrated from [rt.perl.org#58942](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=58942) (status was 'resolved')

Searchable as RT58942$

**Comment by p5pRT** [comment]
### From ambrus@math.bme.hu

#### Created by ambrus@math.bme.hu



Running the following command on my system results in a loop that eats
all available memory fast\.

perl \-we 'grep \{ unpack "a/\(a\)"\, 2 \} 0;'

This happens in both perl\-5\.10\.0 and perl\-5\.8\.0\.

I think the command should terminate quickly instead\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.10.0:

Configured by ambrus at Tue Dec 18 19:35:54 CET 2007.

Summary of my perl5 (revision 5 version 10 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.6.20.6, archname=i686-linux
    uname='linux king 2.6.20.6 #3 smp tue may 29 09:42:52 cest 2007
i686 gnulinux '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-fno-strict-aliasing -pipe
-I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='4.1.1', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /usr/lib64
    libs=-lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.3.6.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.3.6'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib'

Locally applied patches:



@INC for perl 5.10.0:
    /usr/local/lib/perl5/5.10.0/i686-linux
    /usr/local/lib/perl5/5.10.0
    /usr/local/lib/perl5/site_perl/5.10.0/i686-linux
    /usr/local/lib/perl5/site_perl/5.10.0
    /usr/local/lib/perl5/site_perl/5.8.8
    /usr/local/lib/perl5/site_perl
    .


Environment for perl 5.10.0:
    HOME=/home/ambrus
    LANG (unset)
    LANGUAGE=en_HU:en_US:en_GB:en
    LC_CTYPE=hu_HU
    LD_LIBRARY_PATH=/home/ambrus/local/lib/
    LOGDIR (unset)
    PATH=/home/ambrus/local/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @nwc10

On Tue\, Sep 16\, 2008 at 03&#8203;:02&#8203;:35PM \-0700\, Zsban Ambrus wrote&#8203;:
> Running the following command on my system results in a loop that eats
> all available memory fast\.
> 
> perl \-we 'grep \{ unpack "a/\(a\)"\, 2 \} 0;'
> 
> This happens in both perl\-5\.10\.0 and perl\-5\.8\.0\.
> 
> I think the command should terminate quickly instead\.

Thanks for the bug report\. Yes\, it should terminate quickly\.

The bug was be caused by extra values being put onto perl's internal
data stack\, caused by a bug somewhere in the implementation of the unpack
opcode&#8203;:

perl \-Dst \-we 'grep \{ ' \-e 'unpack "a/\(a\)"\, 2' \-e ' \} 0;' 2>&1 | head \-50

EXECUTING\.\.\.

&nbsp;    =>  
\(\-e&#8203;:0\)  enter
&nbsp;    =>  
\(\-e&#8203;:0\)  nextstate
&nbsp;    =>  
\(\-e&#8203;:2\)  pushmark
&nbsp;    =>  \*  
\(\-e&#8203;:2\)  const\(IV\(0\)\)
&nbsp;    =>  \*  IV\(0\)  
\(\-e&#8203;:2\)  grepstart
&nbsp;    =>  \*  IV\(0\)  \*\*  
\(\-e&#8203;:2\)  const\(PV\("a/\(a\)"\\0\)\)
&nbsp;    =>  \*  IV\(0\)  \*\*  PV\("a/\(a\)"\\0\)  
\(\-e&#8203;:2\)  const\(IV\(2\)\)
&nbsp;    =>  \*  IV\(0\)  \*\*  PV\("a/\(a\)"\\0\)  IV\(2\)  
\(\-e&#8203;:2\)  unpack
&nbsp;    =>  \*  IV\(0\)  \*\*  PV\(""\\0\)  PV\(""\\0\)  
\(\-e&#8203;:2\)  scope
&nbsp;    =>  \*  IV\(0\)  \*\*  PV\(""\\0\)  PV\(""\\0\)  
\(\-e&#8203;:2\)  grepwhile
&nbsp;    =>  \*  IV\(0\)  \*  PV\(""\\0\)  \*  
\(\-e&#8203;:2\)  const\(PV\("a/\(a\)"\\0\)\)
&nbsp;    =>  \*  IV\(0\)  \*  PV\(""\\0\)  \*  PV\("a/\(a\)"\\0\)  
\(\-e&#8203;:2\)  const\(PVIV\("2"\\0\)\)
&nbsp;    =>  \*  IV\(0\)  \*  PV\(""\\0\)  \*  PV\("a/\(a\)"\\0\)  PVIV\("2"\\0\)  
\(\-e&#8203;:2\)  unpack
&nbsp;    =>  \*  IV\(0\)  \*  PV\(""\\0\)  \*  PV\(""\\0\)  PV\(""\\0\)  
\(\-e&#8203;:2\)  scope
&nbsp;    =>  \*  IV\(0\)  \*  PV\(""\\0\)  \*  PV\(""\\0\)  PV\(""\\0\)  
\(\-e&#8203;:2\)  grepwhile
&nbsp;    =>  \*  IV\(0\)  \*  PV\(""\\0\)  PV\(""\\0\)  \*  
\(\-e&#8203;:2\)  const\(PV\("a/\(a\)"\\0\)\)
&nbsp;    =>  \*  IV\(0\)  \*  PV\(""\\0\)  PV\(""\\0\)  \*  PV\("a/\(a\)"\\0\)  
\(\-e&#8203;:2\)  const\(PVIV\("2"\\0\)\)
&nbsp;    =>  \*  IV\(0\)  \*  PV\(""\\0\)  PV\(""\\0\)  \*  PV\("a/\(a\)"\\0\)  PVIV\("2"\\0\)  
\(\-e&#8203;:2\)  unpack
&nbsp;    =>  \*  IV\(0\)  \*  PV\(""\\0\)  PV\(""\\0\)  \*  PV\(""\\0\)  PV\(""\\0\)  
\(\-e&#8203;:2\)  scope
&nbsp;    =>  \*  IV\(0\)  \*  PV\(""\\0\)  PV\(""\\0\)  \*  PV\(""\\0\)  PV\(""\\0\)  
\(\-e&#8203;:2\)  grepwhile
&nbsp;    =>  \*  IV\(0\)  \*  PV\(""\\0\)  PV\(""\\0\)  PV\(""\\0\)  \*  
\(\-e&#8203;:2\)  const\(PV\("a/\(a\)"\\0\)\)
&nbsp;    =>  \*  IV\(0\)  \*  PV\(""\\0\)  PV\(""\\0\)  PV\(""\\0\)  \*  PV\("a/\(a\)"\\0\)  
\(\-e&#8203;:2\)  const\(PVIV\("2"\\0\)\)
&nbsp;    =>  \*  IV\(0\)  \*  PV\(""\\0\)  PV\(""\\0\)  PV\(""\\0\)  \*  PV\("a/\(a\)"\\0\)  PVIV\("2"\\0\)  
\(\-e&#8203;:2\)  unpack
&nbsp;    =>  \*  IV\(0\)  \*  PV\(""\\0\)  PV\(""\\0\)  PV\(""\\0\)  \*  PV\(""\\0\)  PV\(""\\0\)  



It's been fixed recently in the development version of perl\, and the fix will
be in 5\.10\.1 and 5\.8\.9\. 5\.8\.9 should be released within a month\.

Nicholas Clark

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From p5p@spam.wizbit.be

Fixed with Change http&#8203;://public\.activestate\.com/cgi\-bin/perlbrowse/p/
33239

Change 33239 by nicholas@&#8203;nicholas\-bouvard on 2008/02/06 01&#8203;:00&#8203;:43

&nbsp;    in unpack\, \(\) groups in scalar context were still returning a 
list\,
&nbsp;    resulting in garbage on the stack\, which could manifest as a 
SEGV
&nbsp;    \(Bug 50256\)

**Comment by p5pRT** [comment]
p5p@spam.wizbit.be - Status changed from 'open' to 'resolved'
